File: Leetcode/515/515.cpp
/* ************************************************************************
> File Name:     515.cpp
> Author:        Jack Yang
> Created Time:  2023年05月13日 (週六) 16時06分35秒
> Description:   Find Largest Value in Each Tree Row 
 ************************************************************************/
#include<iostream>
#include<vector>
#include<queue>
#include <climits>
using namespace std;

struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int>result;
        if(!root)return result;
        queue<TreeNode *>que;
        que.push(root);
        while(!que.empty()){
           int size=que.size();
           int Max=INT_MIN;
           for(int i=0;i<size;i++){
                TreeNode *node=que.front();
                que.pop();
                Max=max(Max,node->val);
                if(node->left)que.push(node->left);
                if(node->right)que.push(node->right);
           }
           result.push_back(Max);
        }
        return result;
    }
};

int main(){
    TreeNode *root=new TreeNode(1);
    root->left=new TreeNode(3);
    root->right=new TreeNode(2);
    TreeNode *ptr;
    ptr=root->left;
    ptr->left=new TreeNode(5);
    ptr->right=new TreeNode(3);
    ptr=root->right;
    ptr->right=new TreeNode(9);
    vector<int>result;
    Solution sol;
    result=sol.largestValues(root);
    for(int &i:result){
        cout << i << " ";
    }
    return 0;
}

File: Leetcode/515.Find_Largest_Value_in_Each_Tree_Row/515.py
from typing import Optional,List
import queue
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        lt=[]
        if root == None:
            return lt
        que=queue.Queue(10001)
        que.put(root)
        while not que.empty():
            size=que.qsize()
            Max=-2147483649
            for i in range(size):
                node=que.get()
                Max=max(Max,node.val)
                if node.left:
                    que.put(node.left)
                if node.right:
                    que.put(node.right)
            lt.append(Max)
        return lt

root=TreeNode(1)
root.left=TreeNode(3)
root.right=TreeNode(2)
ptr=root.left
ptr.left=TreeNode(5)
ptr.right=TreeNode(3)
ptr=root.right
ptr.right=TreeNode(9)
sol=Solution()
result=sol.largestValues(root)
for i in result:
    print(i)
